### PR TITLE
Fix `torch.nn.functional.grid_sample` crashes if `grid` has NaNs

### LINKED ISSUE #42616: Segfault in `torch.nn.functional.grid_sample` in reflection mode
## To Reproduce

~~~python
import torch
input = torch.zeros([1,1,1,3])  # shape [1,1,1,3]
grid = torch.tensor([[[[ 2.5e+38, 0]]]])  # shape  [1, 1, 1, 2]
torch.nn.functional.grid_sample(input=input, grid=grid, padding_mode='reflection')
~~~

This bug only occurs when `padding_mode` is 'reflection'.

## Expected behavior

Expect no segfault

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.4 LTS
GCC version: Could not collect
CMake version: version 3.14.0

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.5.0
[pip] torchvision==0.6.0a0+82fd1c8
[conda] blas 1.0 mkl
[conda] cudatoolkit 10.1.243 h6bb024c_0
[conda] mkl 2020.1 217
[conda] mkl-include 2020.1 217
[conda] mkl-service 2.3.0 py37he904b0f_0
[conda] mkl_fft 1.0.15 py37ha843d7b_0
[conda] mkl_random 1.1.0 py37hd6b4f25_0
[conda] numpy 1.18.1 py37h4f9e942_0
[conda] numpy-base 1.18.1 py37hde5b4d6_1
[conda] pytorch 1.5.0 py3.7_cuda10.1.243_cudnn7.6.3_0 pytorch
[conda] torchvision 0.6.0 py37_cu101 pytorch

## Additional context

cc @ezyang @gchanan @zou3519 @albanD @mruberry

### PR DESCRIPTION
**Pull request by malfet**
In `clip_coordinates` replace `minimum(maximum(in))` composition with `clamp_max(clamp_min(in))` 
Swap order of `clamp_min` operands to clamp NaNs in grid to 0

Fixes #42616

### CODE REVIEW

**Comment by ezyang** [comment]
cc @emcastillo @ngimel

**Review by ngimel — APPROVED** [review]

**Comment by SsnL** [comment]
Wait. Isn't NaN propagation what we are supposed to do?

**Comment by ngimel** [comment]
`clamp` has to propagate nans, so we can't change it. I'm not sure what is expected behavior of `grid_sample` with `nans` in the grid - is it RuntimeError or silently ignoring them?

**Comment by SsnL** [comment]
I thought we decided to propagate NaN as much as we could. I suppose it is more important in grid_sample, which is often used as a layer/op in a neural network. Yet this PR stops propagation IIRC.

**Comment by ngimel** [comment]
But if there are `nan`s in the grid you cannot reasonably propagate them, you can only runtimeError?

**Comment by SsnL** [comment]
Why can't we? If there is `nan` for a location in the grid, shouldn't the output for that location just be `nan`?

**Comment by malfet** [comment]
Propagating NaNs from grid into output would be quite expensive and before this patch would almost certainly result in segfault. 
It's easy to add this logic, but it will unnecessarily slow down the execution of the operator in normal cases.

**Comment by SsnL** [comment]
Hmm sure. Maybe worth saying so in comments then? The previous commit says something like "it is important to clamp NaN", which was unintuitive to me.

File: aten/src/ATen/native/cpu/GridSamplerKernel.cpp [hunk 1]
```diff
@@ -203,7 +203,8 @@ struct ComputeLocationBase<scalar_t, /*align_corners=*/true> {
   }
 
   inline Vec clip_coordinates(const Vec &in) const {
-    return minimum(Vec(max_val), maximum(in, Vec(0)));
+    // Invert order of clamp_min operands in order to clamp Nans to zero
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/GridSamplerKernel.cpp:206]: nice!

**Review by ngimel — COMMENTED** [review]
Can you please clarify in the docs what happens for nan/inf values in the grid? Interestingly original issue was about very large value turning into inf, looks like it should have been handled before by `minimum`/`maximum` clipping?
This fix is for CPU only, what happens on CUDA?

**Comment by malfet** [comment]
> Can you please clarify in the docs what happens for nan/inf values in the grid? 

Done

> Interestingly original issue was about very large value turning into inf, looks like it should have been handled before by `minimum`/`maximum` clipping?

It was turning into `nan` in reflection mode, because `inf` - `inf` == `nan`.

> This fix is for CPU only, what happens on CUDA?

Somehow it's already working like that on CUDA (perhaps GPU automatically turns `nan` into `0` when casting it to int). Extended the test to cover that functionality

**Comment by ngimel** [comment]
Great, can you add that to docs note too, that large values can be converted to `nan` due to overflows during internal operations on the `grid` and subsequent `inf-inf` ?

**Comment by xwang233** [comment]
FYI, the cuda fix is at https://github.com/pytorch/pytorch/pull/35506.

Perhaps we can merge the test in this PR with this and remove `onlyCUDA`?

https://github.com/pytorch/pytorch/blob/3cf2551f2fc8c4d9b53e249d91d40944e1ec8f1c/test/test_nn.py#L10098-L10139

(`issue_24823_1(torch.half)` can be removed since arange half is not supported on cpu)

**Pull request closed without merging**